2015/proposals/schamnad-sanjeed.md

```diff
@@ -0,0 +1,115 @@
+
+## Technical Details
+juliaTLS will be based on the OpenSSL project and will use libopenssl as the external library. The objective of the project is to create the necessary bindings between libopenssl and Julia, and to create a set of functions that will demonstrate the robustness of the implementation. These functions will, minimally, allow the user to
+a. Create and manage private keys, public keys and parameters;
+b. Create and negotiate X.509 certificates, with complete certificate validation; and
+c. Encrypt and Decrypt with the negotiated cypher suite.
+
+juliaTLS will wrap OpenSSL into a format for other Julia packages to use easily. It will incorporate the required functions of OpenSSL for the establishment and teardown of secure socket-based communications.
+
+juliaTLS will consist of three sub-parts: 
+1. The Cryptographic Backend
+This protocol will contain the encryption algorithms used in the Record layer. Its purpose is to encrypt and authenticate packages. 
+The encryption algorithms will be based upon the libcrypto library of OpenSSL. Some features such as Key cryptography and a few hash functions have already been implemented in Julia. Using the �:ccall� functionality of Julia and leveraging, where appropriate, existing work in this area, the following encryption packages will be incorporated into juliaTLS: 
+* Symmetric Ciphers
+o RC4,RC5, 3DES, AES. 
+* Public key Cryptography and matching 
+o RSA, DSA
+* Hash functions and authentication codes
+o MD2, MD5, SHA, SHA1, SHA224, SHA256
+
+Due to the extreme sensitivity of cryptographic implementation, this project will not be developing native implementations of cryptosystems; instead, we will rely on the well-vetted and widely-implemented versions present in the OpenSSL package.
+
+2) The Session Establishment Framework
+The session establishment framework will be responsible for the cipher negotiation, the initial key exchange and any renegotiation functionality, validation of certificates, and the establishment (and teardown) of the TLS session. A few functions for which wrappers for the OpenSSL library(openssl/ssl.h
+) need to be implemented are listed as follows:
+* int TLS_accept(SSL *ssl);
+* int TLS_connect(SSL *ssl);
+* int TLS_read(SSL *ssl, void *buf, int num);
+* int TLS_write(SSL *ssl, const void *buf, int num);
+
+The output of this framework will be an encrypted stream.
+
+Clients and Peers will be authenticated using the Certificate authentication methodology, in particular by using X.509 certificates. This will be based on the methodology incorporated by OpenSSL, in particular the following functions:
+* openssl req -new -key privkey.pem -out cert.csr
+* int SSL_CTX_use_certificate(SSL_CTX *ctx, X509 *x);
+* X509 *TLS_CTX_get0_certificate(const SSL_CTX *ctx);
+* X509 *TLS_get_certificate(const SSL *ssl);
+
+3) The Application Programming Interface (API) layer
```

Comment:
I believe that other languages, e.g. Python or Ruby, has API for OpenSSL/TLS. Did you check it for inspiration?